File: 6/2.py
original_grid = [list(input().strip()) for _ in range(130)]

dirs = {
    "^": (-1,0),
    ">": (0,1),
    "<": (0,-1),
    "v": (1,0)
}

next_dir = {
    "^": ">",
    ">": "v",
    "<": "^",
    "v": "<"
}

def attempt(i, j):
    
    grid = []
    for row in original_grid:
        grid.append(row.copy())

    if grid[i][j] == "^":
        return False
    
    if grid[i][j] == "#":
        return False
    
    grid[i][j] = "#"
    
    curr = "^"

    r = 59
    c = 62

    cache = set()
    
    while True:

        if (r, c, curr) in cache:
            return True
        
        cache.add((r, c, curr))

        x, y = dirs[curr]

        r += x
        c += y

        if r >= 130 or r < 0 or c >= 130 or c < 0:
            return False

        next_tile = grid[r][c]

        if next_tile != "#":
            continue

        r -= x
        c -= y

        curr = next_dir[curr]


total = 0

for i in range(130):
    for j in range(130):
        if attempt(i, j):
            total += 1

print(total)

File: 6/2/2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Move struct {
	r int
	c int
	d string
}

var dirs = map[string][]int{
	"^": {-1, 0},
	">": {0, 1},
	"<": {0, -1},
	"v": {1, 0},
}

var nextDir = map[string]string{
	"^": ">",
	">": "v",
	"<": "^",
	"v": "<",
}

var originalGrid = [130][130]string{}

func attempt(i, j int) bool {
	if originalGrid[i][j] == "^" || originalGrid[i][j] == "#" {
		return false
	}

	grid := [130][130]string{}
	for i, row := range originalGrid {
		for j := range row {
			grid[i][j] = originalGrid[i][j]
		}
	}

	grid[i][j] = "#"

	curr := "^"

	r := 59
	c := 62

	cache := map[Move]bool{}

	for {

		move := Move{r, c, curr}

		if _, ok := cache[move]; ok {
			return true
		}

		cache[move] = true

		val, _ := dirs[curr]

		x, y := val[0], val[1]
		r += x
		c += y

		if r >= 130 || r < 0 || c >= 130 || c < 0 {
			return false
		}

		nextTile := grid[r][c]

		if nextTile != "#" {
			continue
		}

		r -= x
		c -= y

		curr, _ = nextDir[curr]

	}
}

func main() {

	scanner := bufio.NewScanner(os.Stdin)

	i := 0
	for scanner.Scan() {

		line := scanner.Text()
		for j, val := range strings.Split(line, "") {
			originalGrid[i][j] = val
		}
		i++
	}

	total := 0

	for i := 0; i < 130; i++ {
		for j := 0; j < 130; j++ {
			if attempt(i, j) {
				total++
			}
		}
	}

	fmt.Println(total)

}
